cap max column sizes for mssql and oracle

willus10245 opened the pull request:
fixes #9302 

Oracle and MSSQL can return column sizes of upwards of 4GB, which can cause memory allocation errors when trying to allocate that much buffer space. This change keeps the prior fix for Postgres compatibility, but caps large column sizes for Oracle and MSSQL databases.

willus10245 committed:
cap max column sizes for mssql and oracle

Oracle and MSSQL can return column sizes of upwards of 4GB, which can cause memory allocation errors when trying to allocate that much buffer space.
This change keeps the prior fix for Postgres compatibility, but caps large column sizes for Oracle and MSSQL databases.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/erlang/otp?pullRequest=10297) <br/>All committers have signed the CLA.

willus10245 commented:
Got it. Thank you for the thoughtful response! I appreciate the thorough consideration. 

The change in #7000  is what I am attempting to fix. You said that change was reverted? But the change introduced there is still on master.

IngelaAndin commented:
@willus10245  Alas "kvakvs"  stopped working for as and will not have time to look into a better solution. 

Any way this code is really old and a little "AI- googling" results in the following ...
 
"In SQL Server, varchar(8000) is the maximum size for in-row storage, while varchar(max) allows up to 2GB of data, often requiring specific handling in ODBC when exceeding 8000 characters. Using varchar(max) enables storing over 8000 characters, but data exceeding this typically shifts from in-row to out-of-row storage. "

... which can explain the current max value.  Maybe a small improvement could be to  make the max value configurable.  A little more work could perhaps make the code consider varchar(max). We will of course accept things that are clearly an improvement as long as it remains a supported application.

willus10245 committed:
odbc: Cap long-column buffer size to prevent memory exhaustion

SQLDescribeCol can report 0 or a very large display size for
SQL_LONGVARCHAR/SQL_LONGVARBINARY/SQL_WLONGVARCHAR types (e.g.
MSSQL varchar(max) reports 2 GB). Using that value directly for
SQLBindCol buffers makes safe_malloc fail, terminating the port
with memory_allocation_failed.

Clamp the effective size to a configurable limit (default 8001
bytes, preserving prior behavior). Add connect option
max_long_column_size with a hard cap at 256 MiB.

The OPEN_CONNECTION wire protocol is extended with a version
marker (0xFF 0x01) followed by a 4-byte big-endian size value.
Legacy clients without the marker still work; environment variable
ERL_ODBC_MAX_LONG_COLUMN_SIZE provides a fallback for legacy
open messages.

Closes: GH-9302

willus10245 committed:
odbc: Fix memory allocation failure on huge long-column sizes
SQLDescribeCol can report 0 or a very large display size for
SQL_LONGVARCHAR and related types (e.g. MSSQL varchar(max)). Using
that value directly for SQLBindCol buffers can make safe_malloc fail
and terminate the port with memory_allocation_failed (see GH-9302),
regressing the intent of GH-7000.
Clamp the effective size to a default limit (8001 bytes, preserving
prior behavior for drivers that report modest sizes such as
PostgreSQL), with a hard upper bound to avoid runaway allocation.
Add connect option max_long_column_size and an extended OPEN_CONNECTION
wire layout (0xFF 0x01 prefix + 32-bit big-endian size). Legacy
open messages still work; ERL_ODBC_MAX_LONG_COLUMN_SIZE applies when
the extended layout is not used.

IngelaAndin commented:
@willus10245 We will revaluate this PR after next release candidate that has already created a code freez. Although we will deprecate the odbc application we in OTP-29 we want it to be in as good shape as possible and there is an interest from the community to make it  an  hex-package.

IngelaAndin commented:
@willus10245  you need to fix these for testing to be able to proceed. It is probably not your fault in the first place but it is a new check to improve.

```
Error: Whitespace errors detected in modified lines.
Error: Use only spaces for indentation.
lib/odbc/c_src/odbcserver.c:1398: tab in indent.
+	if(sql_type == SQL_LONGVARCHAR || sql_type == SQL_LONGVARBINARY || sql_type == SQL_WLONGVARCHAR) {
lib/odbc/c_src/odbcserver.c:1399: tab in indent.
+	    if(size == 0) {
lib/odbc/c_src/odbcserver.c:1400: tab in indent.
+		size = max_long_column_size(state);
lib/odbc/c_src/odbcserver.c:1401: tab in indent.
+	    } else if(size > max_long_column_size(state)) {
lib/odbc/c_src/odbcserver.c:1402: tab in indent.
+		size = max_long_column_size(state);
lib/odbc/c_src/odbcserver.c:1403: tab in indent.
+	    }
lib/odbc/c_src/odbcserver.c:1404: tab in indent.
+	}
Error: Process completed with exit code 1.
```

IngelaAndin committed:
odbc: Add test for max_long_column_size option

Verify that:
- Valid values (0, positive integers) are accepted
- Invalid values (negative, non-integer) are rejected
- A TEXT column with large data does not crash the port
  when max_long_column_size is configured appropriately

IngelaAndin committed:
odbc: Document truncation behavior for max_long_column_size

Add an explicit warning that data exceeding the configured
max_long_column_size will be silently truncated. Users storing
large values in TEXT/VARCHAR(MAX) columns must increase the
setting to avoid data loss.

IngelaAndin reviewed (dismissed):
I think this is a good enough solution for now. Hopefully this application will be picked up and added to Hex and people using it could make further improvements.
